### PR TITLE
chromium: Set LIBCPLUSPLUS to use libc++

### PR DESCRIPTION
**Pull request by kraj**
This should fix issue seen with
https://github.com/OSSystems/meta-browser/pull/551

Signed-off-by: Khem Raj <raj.khem@gmail.com>

### CODE REVIEW

**Review by rakuco — COMMENTED** [review]
Does this also work with older branches like hardknott? In case of master, where we didn't seem to be passing any toolchain-related arguments per #551, doesn't it mean you also need to pass `-rtlib` and `--unwindlib`?

**Comment by rakuco** [comment]
cc @shr-project @schnitzeltony

**Comment by kraj** [comment]
> Does this also work with older branches like hardknott? In case of master, where we didn't seem to be passing any toolchain-related arguments per #551, doesn't it mean you also need to pass `-rtlib` and `--unwindlib`?

yes it will be a nop on hardknott and dunfell.

**Comment by rakuco** [comment]
> In case of master, where we didn't seem to be passing any toolchain-related arguments per #551, doesn't it mean you also need to pass `-rtlib` and `--unwindlib`?

How about this part?

**Comment by shr-project** [comment]
This didn't work for me, because chromium doesn't seem respect CXXFLAGS/LDFLAGS ("stdlib=libc++" doesn't appear anywhere in log.do_compile - even for other modules than sql).

If I add LIBCPLUSPLUS to TUNE_CCARGS (sort of like it was in meta-clang/hardknott), then it builds OK.

File: meta-chromium/recipes-browser/chromium/chromium.inc [hunk 1]
```diff
@@ -24,6 +24,13 @@ TOOLCHAIN:class-native = "clang"
 # https://github.com/kraj/meta-clang/issues/449).
 RUNTIME = "llvm"
 
+# meta-clang uses RUNTIME variable to add dependencies but does not add needed
+# compiler options since it expects compiler buildtime defaults to be chosen
+# based on RUNTIME variable which defaults to 'gnu' which means we have to
+# ensure right cxxflags and ldflags are set too for chromium irrespective of
+# disto defaults which could be 'gnu' or 'llvm'
+LIBCPLUSPLUS = "-stdlib=libc++"
```
Comment (schnitzeltony) [inline meta-chromium/recipes-browser/chromium/chromium.inc:32]: Sorry but from comment I don't understand what's going on.
Isn't it much simpler:
RUNTIME takes care of additional DEPENDS but does not touch cxx-/ldflags
=> To force libc++ we have to set LIBCPLUSPLUS

Comment (shr-project) [reply meta-chromium/recipes-browser/chromium/chromium.inc:32]: I misunderstood it first as well, but I think Khem talks about the default RUNTIME the cross toolchain defaults to (not as RUNTIME oe variable). Because RUNTIME is already set to llvm in chromium-x11, but that doesn't help with this issue as it doesn't affect anything since https://github.com/kraj/meta-clang/commit/dc5e04e5819a5e7384ae48f1740ac82970a80b8f and https://github.com/kraj/meta-clang/commit/4d066667c3cbba3b88d2f8d6c5640f5f793d10b4 which removed LIBCPLUSPLUS from TUNE_CCARGS as well.

Comment (kraj) [reply meta-chromium/recipes-browser/chromium/chromium.inc:32]: yeah most packages do respect CPPFLAGS/LDFLAGS but here in chromium we might be at odds. Perhaps adding `-stlib=xxx` option can be added to both link and compile stages without issues so perhaps we are ok there to use is via TUNE_CCARGS, let me fix the patch to use TUNE_CCARGS here and also improve the comment a bit

Comment (kraj) [reply meta-chromium/recipes-browser/chromium/chromium.inc:32]: > Sorry but from comment I don't understand what's going on.
> Isn't it much simpler:
> RUNTIME takes care of additional DEPENDS but does not touch cxx-/ldflags
> => To force libc++ we have to set LIBCPLUSPLUS

its not easily possible to allow switching runtime at recipe level, so I would like folks to use it at distro level unless
packages do the right thing to use one or other like chromium, thats why the change is done in meta-clang because in many recipes you may choose LLVM runtime but then it might bring in gnu runtime into build via other dependencies and conflict, then we will end up fixing them which is unwanted work.

**Comment by shr-project** [comment]
Last version works for me +1

**Comment by schnitzeltony** [comment]
Builds fine - thanks

File: meta-chromium/recipes-browser/chromium/chromium.inc [hunk 2]
```diff
@@ -24,6 +24,22 @@ TOOLCHAIN:class-native = "clang"
...
+# global RUNTIME variable setting, which defaults to 'gnu'
+# For chromium libc++ is expected to be used so, irrespective of
+# ensure right build flags are set too for chromium irrespective of
+# disto defaults which could be 'gnu' or 'llvm', here we need to set both
+# target and buid flags since chromium builds native pieces along with target
+# pieces and does not have separate native variant to provide the native tools
+# this also ensures that it does not rely too much on meta-clang's defaults
+LIBCPLUSPLUS = "-stdlib=libc++"
+
+# LIBCPLUSPLUS is included in CXXFLAGS and LDFLAGS, but chromium doesn't respect it
```
Comment (rakuco) [inline meta-chromium/recipes-browser/chromium/chromium.inc:39]: Hmm, does it really not respect those variables? The logic in [gn-utils.inc's write_toolchain_file()](https://github.com/OSSystems/meta-browser/blob/3405e3af0fa67a82bd28884cac47314bf847e6b7/meta-chromium/recipes-browser/chromium/gn-utils.inc#L49) reads all the {BUILD,TARGET}_{CPP,CXX,LD,CC}FLAGS, so I'm wondering if there's something to fix there instead.

Comment (shr-project) [reply meta-chromium/recipes-browser/chromium/chromium.inc:39]: CXXFLAGS != TARGET_CXXFLAGS (LIBCPLUSPLUS is appended to CXXFLAGS but not TARGET_CXXFLAGS).

```
# $CXXFLAGS [7 operations]                                                                                                                                                   |
#   :append[pn-xdelta3:toolchain-clang] /jenkins/mjansa/build/nodistro/honister/meta-clang/conf/nonclangable.conf:273                                                        |
#     " -Wno-error=reserved-user-defined-literal"                                                                                                                            |
#   :remove[pn-mozjs:toolchain-clang] /jenkins/mjansa/build/nodistro/honister/meta-clang/conf/nonclangable.conf:275                                                          |
#     "-fno-tree-vrp"                                                                                                                                                        |
#   :append[pn-apt:toolchain-clang] /jenkins/mjansa/build/nodistro/honister/meta-clang/conf/nonclangable.conf:353                                                            |
#     " -Wno-c++11-narrowing"                                                                                                                                                |
#   exported /jenkins/mjansa/build/nodistro/honister/oe-core/meta/conf/bitbake.conf:579                                                                                      |
#     [export] "1"                                                                                                                                                           |
#   set /jenkins/mjansa/build/nodistro/honister/oe-core/meta/conf/bitbake.conf:579                                                                                           |
#     "${TARGET_CXXFLAGS}"                                                                                                                                                   |
#   append /jenkins/mjansa/build/nodistro/honister/oe-core/meta/conf/distro/include/tclibc-glibc.inc:17                                                                      |
#     "-fvisibility-inlines-hidden"                                                                                                                                          |
#   :append[toolchain-clang] /jenkins/mjansa/build/nodistro/honister/meta-clang/classes/clang.bbclass:29                                                                     |
#     " ${LIBCPLUSPLUS}"                                                                                                                                                     |
# pre-expansion value:                                                                                                                                                       |
#   "${TARGET_CXXFLAGS} -fvisibility-inlines-hidden ${LIBCPLUSPLUS}"                                                                                                         |
export CXXFLAGS=" -O2 -pipe  -feliminate-unused-debug-types -fmacro-prefix-map=/jenkins/mjansa/build/nodistro/honister/BUILD/work/core2-32-oe-linux/chromium-x11/92.0.4515.15|
9-r0=/usr/src/debug/chromium-x11/92.0.4515.159-r0                      -fdebug-prefix-map=/jenkins/mjansa/build/nodistro/honister/BUILD/work/core2-32-oe-linux/chromium-x11/9|
2.0.4515.159-r0=/usr/src/debug/chromium-x11/92.0.4515.159-r0                      -fdebug-prefix-map=/jenkins/mjansa/build/nodistro/honister/BUILD/work/core2-32-oe-linux/chr|
omium-x11/92.0.4515.159-r0/recipe-sysroot=                      -fdebug-prefix-map=/jenkins/mjansa/build/nodistro/honister/BUILD/work/core2-32-oe-linux/chromium-x11/92.0.451|
5.159-r0/recipe-sysroot-native= -g1 -fvisibility-inlines-hidden -stdlib=libc++"  
```

Changing:
```
meta-chromium/recipes-browser/chromium/gn-utils.inc:        'extra_cxxflags': d.expand('${TARGET_CXXFLAGS}'),
```
To CXXFLAGS should work.

Comment (rakuco) [reply meta-chromium/recipes-browser/chromium/chromium.inc:39]: > CXXFLAGS != TARGET_CXXFLAGS (LIBCPLUSPLUS is appended to CXXFLAGS but not TARGET_CXXFLAGS).

Ah, I see, thanks for the clarification. I wonder if LIBCPLUSPLUS shouldn't be added to TARGET_CXXFLAGS instead? According to the docs CXXFLAGS is set as an environment variable in the compilation step; I think if it's set during the configuration step it should work with GN as well, but if we can capture all necessary flags when write_toolchain_file() is set it's even better to avoid discrepancies and flags showing up out of nowhere.

Comment (shr-project) [reply meta-chromium/recipes-browser/chromium/chromium.inc:39]: Using variables without TARGET_ prefix is correct and e.g. cmake is using the same when generating its toolchain config:
https://git.openembedded.org/openembedded-core/tree/meta/classes/cmake.bbclass#n28

I've verified that LIBCPLUSPLUS works with: https://github.com/shr-project/meta-browser/commit/1525aa425df2fce4ad7a98f8b3e2b8716fff6b08

**Comment by shr-project** [comment]
LGTM

File: meta-chromium/recipes-browser/chromium/chromium.inc [hunk 3]
```diff
@@ -24,6 +24,18 @@ TOOLCHAIN:class-native = "clang"
 # https://github.com/kraj/meta-clang/issues/449).
 RUNTIME = "llvm"
 
+# meta-clang uses RUNTIME variable to add build dependencies on libcxx/compiler-rt
+# but does not add compiler options to commandline since it expects compiler's own
+# defaults to be used and compiler's buildtime defaults are chosen based on
+# global RUNTIME variable setting, which defaults to 'gnu'
+# For chromium libc++ is expected to be used so, irrespective of
```
Comment (rakuco) [inline meta-chromium/recipes-browser/chromium/chromium.inc:31]: The first ", irrespective of" seems like a leftover from a previous version of this comment.

File: meta-chromium/recipes-browser/chromium/chromium.inc [hunk 4]
```diff
@@ -24,6 +24,18 @@ TOOLCHAIN:class-native = "clang"
...
+# ensure right build flags are set too for chromium irrespective of
+# disto defaults which could be 'gnu' or 'llvm', here we need to set both
```
Comment (rakuco) [inline meta-chromium/recipes-browser/chromium/chromium.inc:33]: disto -> distro
I'd also add a full stop before "here", otherwise you end up with a very long paragraph with only one sentence that's hard to follow.

File: meta-chromium/recipes-browser/chromium/chromium.inc [hunk 5]
```diff
@@ -24,6 +24,18 @@ TOOLCHAIN:class-native = "clang"
...
+# target and buid flags since chromium builds native pieces along with target
```
Comment (rakuco) [inline meta-chromium/recipes-browser/chromium/chromium.inc:34]: buid -> build

**Review by rakuco — COMMENTED** [review]
This looks a lot better, nice!

One thing: do we still need to continue setting `RUNTIME = "llvm"`? I guess we need it at least for older branches, but I'm not sure.

**Review by rakuco — COMMENTED** [review]
A few remaining questions:
* Do we still need to set `RUNTIME="llvm"`?
* Is this enough to make the build use compiler-rt and the rest of the LLVM toolchain in terms of arguments passed to the compiler/linker?

**Comment by kraj** [comment]
> A few remaining questions:
> 
> * Do we still need to set `RUNTIME="llvm"`?

yes since it will add needed DEPENDs correctly.

> * Is this enough to make the build use compiler-rt and the rest of the LLVM toolchain in terms of arguments passed to the compiler/linker?

yes what we have is enough for now.

**Review by rakuco — APPROVED** [review]
lgtm, thank you both for working on this

**Comment by rakuco** [comment]
@kraj @schnitzeltony I have to revert @kraj's commit, it broke the hardknott build for me. Setting `LIBCPLUSPLUS` actually started preventing the libc++ headers from being installed into `recipe-sysroot/usr/include`.
